### PR TITLE
Implemented the ECP part and added several missing factors in the TREXIO interface.

### PR DESCRIPTION
**Pull request by kousuke-nakano**
@matthew-hennefarth 

Sorry for my late response. Today, I have implemented lines to write ECP information to a TREXIO file. Please have a look at my pull request.

The ECP part works. However, a problem I found is that the `cc-pVXZ` basis sets accompanied with `ccECP`, which are often used for QMC calculations, did not pass the following assertion line in `trexio.py`.

>       assert all(mol._bas[:,gto.NCTR_OF] == 1)

Would you give me a clue for solving this problem?

#60

### CODE REVIEW

**Comment by matthew-hennefarth** [comment]
> The ECP part works. However, a problem I found is that the `cc-pVXZ` basis sets accompanied with `ccECP`, which are often used for QMC calculations, did not pass the following assertion line in `trexio.py`.
> 
> ```
>   assert all(mol._bas[:,gto.NCTR_OF] == 1)
> ```

It seems that `mol._bas` is the internal representation for the libcint library, and `NCTR_OF` is a constant being 3 (which I believe represents number of cartesians; ie x, y, z). Unfortunately, I do not know what the internal representation of `_bas` is supposed to be.

Perhaps, since @sunqm wrote the assert statement and knows the details of libcint, he can provide you with more guidance.

**Comment by kousuke-nakano** [comment]
@matthew-hennefarth, thank you very much. I will wait for @sunqm's reply.

**Comment by kousuke-nakano** [comment]
@matthew-hennefarth 

I found that `mol._bas[:,gto.NCTR_OF]` refers to the numbers of generalized contractions (i.e., one set of exponents and multiple sets of coefficients, like molpro and molcas). See this [issue](https://github.com/TREX-CoE/trexio_tools/issues/42#issuecomment-2463880713). Do you know how to switch it off with some option when we build a `mol`?

**Comment by matthew-hennefarth** [comment]
> I found that `mol._bas[:,gto.NCTR_OF]` refers to the numbers of generalized contractions (i.e., one set of exponents and multiple sets of coefficients, like molpro and molcas). See this [issue](https://github.com/TREX-CoE/trexio_tools/issues/42#issuecomment-2463880713). Do you know how to switch it off with some option when we build a `mol`?

I am not sure how to turn it off, sorry. From the discussion you linked it seems like the best bet would be to just manually expand the contraction set.

**Comment by kousuke-nakano** [comment]
Dear @matthew-hennefarth, Sure. Thank you! Anyway, this is nothing with the ECP part. 

I have completed implementing a function to *write* the ECP part to a `TREXIO` file. The *read* function is my ToDO work. In addition, I have added several missing factors, such as `AO_normalization`, see #80.

File: pyscf/tools/trexio.py [hunk 1]
```diff
@@ -27,15 +29,39 @@ def to_trexio(obj, filename, backend='h5'):
...
     trexio.write_nucleus_label(trexio_file, labels)
-    if mol._ecp:
-        raise NotImplementedError
-    trexio.write_nucleus_charge(trexio_file, mol.atom_charges())
-    trexio.write_nucleus_coord(trexio_file, mol.atom_coords())
+    if mol.symmetry:
+        trexio.write_nucleus_point_group(trexio_file, mol.groupname)
 
+    # 2.3 Periodic boundary calculations
+    try:
```
Comment (sunqm) [inline pyscf/tools/trexio.py:48]: If you don't need to consider the compatibility of old pyscf versions, PBC can be checked using isinstance(mol, pbc.gto.Cell) . Starting from pyscf-2.5, gto.Mole and pbc.gto.Cell represent two distinct classes.

Comment (kousuke-nakano) [reply pyscf/tools/trexio.py:48]: Dear @sunqm, Sure, Thank you very much. I am now implementing the PBC part, in which I do this check as you proposed.

**Comment by kousuke-nakano** [comment]
Sorry @sunqm, I committed new changes ([6949296](https://github.com/pyscf/pyscf-forge/pull/68/commits/694929638f2f89892928e5f29a90ea180e39fcfe)) by mistake. I have reverted it ([db9cde9](https://github.com/pyscf/pyscf-forge/pull/68/commits/db9cde9e7c5128f5f118e939ac52d48ca4d9e812)).

**Review by sunqm — APPROVED** [review]

**Comment by kousuke-nakano** [comment]
@sunqm, I have resolved a trivial conflict.

**Comment by kousuke-nakano** [comment]
Dear @sunqm, Thank you for running the workflows. One of the workflows failed at the part unrelated to this pull request.

**Comment by q-posev** [comment]
@kousuke-nakano are you done with this PR? 

I left one comment about the UHF case, let me know if you prefer to fix it here or if you prefer me to do it in another PR. It should not be too complicated, just a question of extracting the number of alpha and beta electrons from the UHF object.

**Comment by kousuke-nakano** [comment]
Dear @q-posev, @sunqm has already implemented both RHF and UHF in #60.

**Comment by q-posev** [comment]
@kousuke-nakano yes, but on line 63 in `trexio.py` in this PR you added the following line:
```python
    electron_up_num, electron_dn_num = mol.nelec
```
which is not true for UHF. 

The original code of @sunqm was not producing the number of electrons.

**Comment by q-posev** [comment]
The electron writing can be probably moved to `_scf_to_trexio` function where we have an if/else block depending on the output of ` isinstance(mf, scf.uhf.UHF)` (I mean [this line](https://github.com/pyscf/pyscf-forge/pull/68/files#diff-ec9a510320bbc0d6f50c9c5364cfa300ee471c2770c84656d4cdf70864e8384dR206)).

**Comment by kousuke-nakano** [comment]
Dear @q-posev,

> @kousuke-nakano yes, but on line 63 in `trexio.py` in this PR you added the following line:
> 
> ```python
>     electron_up_num, electron_dn_num = mol.nelec
> ```
> 
> which is not true for UHF.
> 
> The original code of @sunqm was not producing the number of electrons.

The following code gives the correct numbers of up and down electrons. 

> import pyscf
> 
> mol_S_0 = pyscf.M(atom='H 0 0 0; F 0 0 1', basis='6-31g*', cart=True)
> electron_up_num, electron_dn_num = mol_S_0.nelec
> print(electron_up_num, electron_dn_num) # -> 5 5
> 
> mol_S_2 = pyscf.M(atom='H 0 0 0; F 0 0 1', basis='6-31g*', cart=True)
> mol_S_2.spin = 2
> electron_up_num, electron_dn_num = mol_S_2.nelec
> print(electron_up_num, electron_dn_num) # -> 6 4

**Comment by q-posev** [comment]
@kousuke-nakano you are totally right, my bad! Please ignore my comments above.

**Pull request merged**